test/Microsoft.ML.TestFrameworkCommon/RetryTestCase.cs

```diff
@@ -0,0 +1,120 @@
+using Xunit.Abstractions;
+using Xunit.Sdk;
+
+namespace Microsoft.ML.TestFrameworkCommon
+{
+    [Serializable]
+    public class RetryTestCase : XunitTestCase
+    {
+        private int maxRetries;
+
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        [Obsolete("Called by the de-serializer", true)]
+        public RetryTestCase() { }
+
+        public RetryTestCase(IMessageSink diagnosticMessageSink, TestMethodDisplay testMethodDisplay, 
+            ITestMethod testMethod, int maxRetries)
+            : base(diagnosticMessageSink, testMethodDisplay, TestMethodDisplayOptions.None, testMethod, testMethodArguments: null)
+        {
+            this.maxRetries = maxRetries;
+        }
+
+        // This method is called by the xUnit test framework classes to run the test case. We will do the
+        // loop here, forwarding on to the implementation in XunitTestCase to do the heavy lifting. We will
+        // continue to re-run the test until the aggregator has an error (meaning that some internal error
+        // condition happened), or the test runs without failure, or we've hit the maximum number of tries.
+        public override async Task<RunSummary> RunAsync(IMessageSink diagnosticMessageSink,
+                                                        IMessageBus messageBus,
+                                                        object[] constructorArguments,
+                                                        ExceptionAggregator aggregator,
+                                                        CancellationTokenSource cancellationTokenSource)
+        {
+            var runCount = 0;
+
+            while (true)
+            {
+                // This is really the only tricky bit: we need to capture and delay messages (since those will
+                // contain run status) until we know we've decided to accept the final result;
+                var delayedMessageBus = new DelayedMessageBus(messageBus);
+
+                RunSummary summary = await base.RunAsync(diagnosticMessageSink, delayedMessageBus, constructorArguments, aggregator, cancellationTokenSource);
+                if (aggregator.HasExceptions || summary.Failed > 0)
+                {
+                    var details = ExtractTestFailDetailsFromMessageBus(delayedMessageBus);
+                    var errorMessage = $"Execution of '{DisplayName}' failed (attempt #{runCount + 1}) with details {details}.";
+
+                    diagnosticMessageSink.OnMessage(new DiagnosticMessage(errorMessage));
+                    Console.WriteLine(errorMessage);
+                }
+
+                if (summary.Failed == 0 || ++runCount >= maxRetries)
+                {
+                    delayedMessageBus.Dispose();  // Sends all the delayed messages
+                    return summary;
+                }
+            }
+        }
+
+        private static string ExtractTestFailDetailsFromMessageBus(DelayedMessageBus delayedMessageBus)
```

Comment:
Will the information about the number of times a particular test was retried and the call stack associated with the failure be visible in the test output log on the build server?

Reply:
Yes, I printed this information to Console and this information is visible in test output as well.

Ideally this kind of information can be put into a centralized and we can view what tests are failing easily like what I did before to put these information into a SQL DB but that can be in a separate PR.

---
In reply to: [367724530](https://github.com/dotnet/machinelearning/pull/4667#discussion_r367724530) [](ancestors = 367724530)